File: 213. House Robber II.js
// https://leetcode.com/problems/house-robber-ii/

/**
 * @param {number[]} nums
 * @return {number}
 */
var rob = function(nums) {

    const memo = new Array(nums.length).fill(0).map(() => new Array(2));

    const r = (i, canRobLast) => {
        if(i >= nums.length) return 0;
        if(i === nums.length-1) return canRobLast ? nums[i] : 0;

        if(memo[i][canRobLast] !== undefined) return memo[i][canRobLast];

        if(!i){
            return memo[i][canRobLast] = Math.max(
                nums[i] + r(i+2, false), // rob this house
                0 + r(i+1, true) // do not rob this house
            );
        }

        return memo[i][canRobLast] = Math.max(
            nums[i] + r(i+2, canRobLast), // rob this house
            0 + r(i+1, canRobLast) // do not rob this house
        );
    };

    return r(0, true);
};

File: 213. House Robber II.go
// https://leetcode.com/problems/house-robber-ii/

package leetcode

func rob(nums []int) int {

	memo := make([]int, 2*len(nums))
	for i := range memo {
		memo[i] = -1
	}
	var m func(int, int) int
	m = func(i, canRobLast int) int {
		offset := len(nums) * canRobLast
		return offset + i
	}

	var r func(i, canRobLast int) int
	r = func(i, canRobLast int) int {
		memoIndex := m(i, canRobLast)

		switch {
		case i >= len(nums): // out of bounds. Can't rob anything
			return 0
		case i == len(nums)-1: // last house. Rob if we haven't robbed the first
			return canRobLast * nums[i]
		case memo[memoIndex] != -1: // Use the memo table if we've been here before
			return memo[memoIndex]
		case i == 0: // if we're starting off, decide whether or not to rob the first house
			memo[memoIndex] = max(
				nums[i]+r(i+2, 0), // rob this house
				0+r(i+1, 1),       // do not rob this house
			)
		default: // otherwise just decide whether or not to rob the current house
			memo[memoIndex] = max(
				nums[i]+r(i+2, canRobLast), // rob this house
				0+r(i+1, canRobLast),       // do not rob this house
			)
		}
		return memo[memoIndex]
	}

	return r(0, 0)
}
